File: ios/Sysmon/StatusStore.swift
import Foundation
import UserNotifications

// StatusStore is the single source of truth for live monitoring state.
//
// Rather than each screen fetching /api/monitoring/status on its own timer,
// one poller keeps a host map warm and both the Alerts and Hosts screens
// observe it. The first fetch pulls the full status; every subsequent poll
// asks for only what changed via ?since=<rev>, so a steady-state poll
// transfers a near-empty delta instead of the whole host list.
@MainActor
final class StatusStore: ObservableObject {
    static let shared = StatusStore()

    @Published private(set) var hosts: [Host] = []
    @Published private(set) var stats: Stats?
    @Published private(set) var daemon: DaemonInfo?
    @Published private(set) var error: String?
    @Published private(set) var loading = true

    // Debounced connectivity for the LIVE pill: a single failed poll is
    // routine (a blip in coverage, a dropped socket) and the next poll
    // usually succeeds - flipping on every hiccup would read OFFLINE
    // half the time on a healthy connection. Only consecutive failures
    // count.
    @Published private(set) var offline = false
    private var failStreak = 0
    private let offlineAfterFailures = 3

    // Poll cadence while the app is foregrounded.
    private let pollInterval: UInt64 = 5_000_000_000 // 5s

    private var rev: Int64 = 0
    private var hostIndex: [String: Host] = [:]
    private var pollTask: Task<Void, Never>?

    var alerts: [Host] { hosts.filter { !$0.isOK } }

    // Begin (or resume) polling. Idempotent - safe to call on every
    // scene-active transition and view appearance.
    func start() {
        guard pollTask == nil else { return }
        pollTask = Task { [weak self] in await self?.pollLoop() }
    }

    // Pause polling (app backgrounded). Retains the last known state so the
    // UI doesn't flash empty when we resume.
    func stop() {
        pollTask?.cancel()
        pollTask = nil
    }

    // Force an immediate fetch (pull-to-refresh). Uses the delta path when
    // we already have a revision, so a refresh is just as cheap as a poll.
    func refreshNow() async {
        await fetch()
    }

    private func pollLoop() async {
        await fetch()
        while !Task.isCancelled {
            try? await Task.sleep(nanoseconds: pollInterval)
            if Task.isCancelled { break }
            await fetch()
        }
    }

    private func fetch() async {
        guard let session = Session.shared, let token = session.token,
              !session.serverURL.isEmpty else { return }
        let api = API(baseURL: session.serverURL, token: token)
        do {
            if rev == 0 {
                let status: StatusResponse = try await api.get("/api/monitoring/status")
                applyFull(status.hosts, stats: status.statistics,
                          daemon: status.daemon, rev: status.rev ?? 0)
            } else {
                let delta: StatusDelta =
                    try await api.get("/api/monitoring/status?since=\(rev)")
                applyDelta(delta)
            }
            error = nil
            failStreak = 0
            offline = false
        } catch is CancellationError {
            return // never count a cancelled poll as a network failure
        } catch let e as URLError where e.code == .cancelled {
            return // URLSession's flavor of cancellation (stop() mid-request)
        } catch let e as APIError {
            error = e.message
            noteFailure()
        } catch {
            // `self.` required: a bare catch binds the thrown value to an
            // implicit constant named `error`, shadowing our published var.
            self.error = "Connection failed"
            noteFailure()
        }
        loading = false
        publishAlertCount()
    }

    private func noteFailure() {
        failStreak += 1
        if failStreak >= offlineAfterFailures {
            offline = true
        }
    }

    private func applyFull(_ list: [Host], stats: Stats, daemon: DaemonInfo?, rev: Int64) {
        hostIndex = Dictionary(list.map { ($0.id, $0) }, uniquingKeysWith: { _, b in b })
        self.stats = stats
        self.daemon = daemon
        self.rev = rev
        rebuildHosts()
    }

    private func applyDelta(_ d: StatusDelta) {
        if d.full {
            applyFull(d.changed ?? [], stats: d.statistics, daemon: d.daemon, rev: d.rev)
            return
        }
        for h in d.changed ?? [] { hostIndex[h.id] = h }
        for name in d.removed ?? [] { hostIndex.removeValue(forKey: name) }
        stats = d.statistics
        daemon = d.daemon
        rev = d.rev
        rebuildHosts()
    }

    private func rebuildHosts() {
        // Secondary key: dictionary order is nondeterministic and Swift's
        // sort is unstable, so hosts sharing a hostname (multiple objects
        // on one IP) would visually swap rows between polls without it.
        hosts = hostIndex.values.sorted {
            ($0.hostname, $0.id) < ($1.hostname, $1.id)
        }
    }

    private func publishAlertCount() {
        let count = alerts.count
        Session.shared?.alertCount = count
        Task { try? await UNUserNotificationCenter.current().setBadgeCount(count) }
    }

    // Clear all state on logout so a subsequent login starts from a full
    // fetch rather than deltas against a stale revision.
    func reset() {
        stop()
        rev = 0
        hostIndex = [:]
        hosts = []
        stats = nil
        daemon = nil
        error = nil
        failStreak = 0
        offline = false
        loading = true
    }
}

File: android/app/src/main/java/com/sysmon/app/StatusStore.kt
package com.sysmon.app

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

/**
 * Single source of truth for live monitoring state.
 *
 * Rather than each screen fetching /api/monitoring/status on its own timer,
 * one poller keeps a host map warm and both the Alerts and Hosts screens
 * observe it. The first fetch pulls the full status; every subsequent poll
 * asks for only what changed via ?since=<rev>, so a steady-state poll
 * transfers a near-empty delta instead of the whole host list.
 */
object StatusStore {
    var hosts by mutableStateOf<List<Host>>(emptyList())
        private set
    var stats by mutableStateOf<Stats?>(null)
        private set
    var daemon by mutableStateOf<DaemonInfo?>(null)
        private set
    var error by mutableStateOf<String?>(null)
        private set
    var loading by mutableStateOf(true)
        private set

    // Debounced connectivity for the LIVE pill: a single failed poll is
    // routine (stale keep-alive sockets, a blip in coverage) and the next
    // poll usually succeeds - flipping the pill on every hiccup made it
    // read OFFLINE half the time on a perfectly healthy connection. Only
    // consecutive failures count.
    var offline by mutableStateOf(false)
        private set

    // The server answered but no sysmond is reporting to it. Nothing is
    // being monitored, which "all good" must not be allowed to claim.
    var degraded by mutableStateOf(false)
        private set
    private var failStreak = 0
    private const val OFFLINE_AFTER_FAILURES = 3

    private const val POLL_MS = 5000L

    private var rev = 0L
    private val index = LinkedHashMap<String, Host>()
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private var pollJob: Job? = null

    val alerts: List<Host> get() = hosts.filter { !it.isOK }
    // The active board: what nobody has claimed yet. Acked alerts are
    // triaged, not gone - screens list them separately.
    val unackedAlerts: List<Host> get() = hosts.filter { !it.isOK && !it.acked }
    val ackedAlerts: List<Host> get() = hosts.filter { !it.isOK && it.acked }

    private fun key(h: Host): String = h.objectName.ifEmpty { h.hostname }

    /** Begin (or resume) polling. Idempotent. */
    fun start() {
        if (pollJob?.isActive == true) return
        pollJob = scope.launch { pollLoop() }
    }

    /** Pause polling (app backgrounded); retains last known state. */
    fun stop() {
        pollJob?.cancel()
        pollJob = null
    }

    /** Force an immediate fetch (pull-to-refresh / manual refresh). */
    suspend fun refreshNow() = fetch()

    private suspend fun pollLoop() {
        fetch()
        while (scope.isActive) {
            delay(POLL_MS)
            fetch()
        }
    }

    private suspend fun fetch() {
        if (!Session.isLoggedIn()) return
        try {
            if (rev == 0L) {
                val status = Api.status()
                applyFull(status.hosts, status.statistics, status.daemon, status.rev)
                degraded = status.sitesReachable == 0
            } else {
                val delta = Api.statusDelta(rev)
                applyDelta(delta)
                degraded = delta.sitesReachable == 0
            }
            error = null
            failStreak = 0
            offline = false
        } catch (e: kotlinx.coroutines.CancellationException) {
            throw e // never count a cancelled poll as a network failure
        } catch (e: Exception) {
            error = e.message
            failStreak++
            if (failStreak >= OFFLINE_AFTER_FAILURES) offline = true
        }
        loading = false
        Session.alertCount = unackedAlerts.size
    }

    private fun applyFull(list: List<Host>, s: Stats, d: DaemonInfo, r: Long) {
        index.clear()
        list.forEach { index[key(it)] = it }
        stats = s
        daemon = d
        rev = r
        rebuild()
    }

    private fun applyDelta(d: StatusDelta) {
        if (d.full) {
            applyFull(d.changed, d.statistics, d.daemon, d.rev)
            return
        }
        d.changed.forEach { index[key(it)] = it }
        d.removed.forEach { index.remove(it) }
        stats = d.statistics
        daemon = d.daemon
        rev = d.rev
        rebuild()
    }

    private fun rebuild() {
        hosts = index.values.sortedBy { it.hostname }
    }

    /** Clear all state on logout so the next login starts from a full fetch. */
    fun reset() {
        stop()
        rev = 0L
        index.clear()
        hosts = emptyList()
        stats = null
        daemon = null
        error = null
        failStreak = 0
        offline = false
        degraded = false
        loading = true
    }
}
